iroh-rpc/src/client.rs

```diff
@@ -234,36 +238,93 @@ mod test {
+    use tracing_subscriber::{fmt, prelude::*, EnvFilter};
 
     use crate::behaviour::Behaviour;
     use crate::builder::RpcBuilder;
     use crate::handler;
     use crate::serde::{deserialize_request, serialize_response, Deserialize};
+    use crate::stream::{Header, OutStream, StreamConfig};
     use crate::swarm;
 
     #[derive(Deserialize, Serialize, Debug, Clone)]
-    struct GetRequest {
-        resource_id: String,
+    struct StreamRequest {
+        size: u64,
+        chunk_size: u64,
     }
 
-    #[derive(Deserialize, Serialize, Debug, Clone, PartialEq)]
-    struct GetResponse {
-        response: String,
+    struct TestReader {
+        chunk_size: usize,
+        num_chunks: usize,
+    }
+
+    impl std::io::Read for TestReader {
+        fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
+            if self.num_chunks == 0 {
+                return Ok(0);
+            };
+            let len = self.chunk_size;
+            let new_data = vec![0xf; len];
+            buf[..len].copy_from_slice(&new_data);
+            self.num_chunks -= 1;
+            Ok(len)
+        }
     }
 
     // TODO: next improvement should be that the serialization and deserialization happen
     // for you in the rpc library, rather than having to do it yourself in the handler function
-    async fn get(
-        state: handler::State<String>,
-        _stream_id: Option<u64>,
+    async fn stream_bytes(
+        _state: handler::State<String>,
+        cfg: Option<StreamConfig>,
         param: Vec<u8>,
     ) -> Result<Vec<u8>, RpcError> {
-        let _: GetRequest = deserialize_request(&param)?;
+        let req: StreamRequest = deserialize_request(&param)?;
+        let size = req.size;
+        let chunk_size = req.chunk_size;
+
+        // get config
+        let cfg = match cfg {
+            Some(c) => c,
+            None => return Err(RpcError::NoStreamConfig),
+        };
 
-        let bytes = serialize_response(GetResponse {
-            response: state.clone(),
-        })?;
-        Ok(bytes)
+        // typically this is where you would load the content,
+        // determine the size of the content & specify what
+        // chunk sizes you are sending back
+        // then, construct a header based on these specifications
+        // in this example, we are pulling the size & chunk_size
+        // from the request
+        let header = Header::new(cfg.id, size, chunk_size);
+
+        // construct BufReader from the loaded content
+        let r = TestReader {
+            num_chunks: {
+                let mut num_chunks = 0;
+                if size % chunk_size != 0 {
+                    num_chunks += 1;
+                };
+                num_chunks + (size / chunk_size) as usize
+            },
+            chunk_size: chunk_size as usize,
+        };
+        let r = std::io::BufReader::new(r);
+
+        // construct stream
+        let mut stream = OutStream::new(cfg, header.clone(), Box::new(r));
```

Comment:
This needs to be an `AsyncRead` thing, or potentially we will want this to be an async channel. But our io will be async, and this forces the use of sync io now into our pipeline.

(We can defer this, but need to fix this when we go and use it)

Reply:
Is this because of the `Read/BufReader` requirement in `OutStream`?
I'll leave a note.